### PR TITLE
E2e pw/migrate api core tests to use playwright 34126

### PR DESCRIPTION
**Pull request by nigeljamesstevenson**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

This PR is for the creation of api-core-tests in playwright. The following test groups were created:
coupons - closes 353-gh-woocommerce/woocommerce-quality
hello - closes 352-gh-woocommerce/woocommerce-quality
orders - closes 354-gh-woocommerce/woocommerce-quality
products - closes 355-gh-woocommerce/woocommerce-quality
refunds - closes 356-gh-woocommerce/woocommerce-quality
shipping - closes 357-gh-woocommerce/woocommerce-quality

### How to test the changes in this Pull Request:

1. Ensure test environment has been initialised correctly
2. Navigate to woocommerce/plugins/woocommerce
3. USE_WP_ENV=1
4. pnpm playwright test --config=tests/api-core-tests/playwright.config.js
5. 120 tests should execute successfully

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

**Comment by zhongruige** [comment]
Thanks for getting this in @nigeljamesstevenson!

When I went to test this locally, and I wasn't able to get the tests to pass. I kept hitting an invalid username error (which resulted in a 500 returned status for the tests):

```
{
  "code": "invalid_username",
  "message": "<strong>Error</strong>: The username <strong>undefined</strong> is not registered on this site. If you are unsure of your username, try your email address instead.",
  "data": null
}
```

I am not sure if this was the way I spun up the test environment. I did the following: in `plugins/woocommerce` I ran `pnpm docker:up` (I wasn't able to get `pnpm docker:up --filter=woocommerce` to work for some reason, but this did).

Is there any other configuration that needs to happen (such as a local `.env` file or something similar) to set things up? Also wanted to check in how you were spinning up and initializing the test environment just to make sure I didn't miss a step, either.

Thanks!

**Comment by zhongruige** [comment]
Also as a side note, we may want to add a `gitignore` around the output file:

```
git checkout trunk
error: Your local changes to the following files would be overwritten by checkout:
        plugins/woocommerce/tests/api-core-tests/output/index.html
Please commit your changes or stash them before you switch branches.
```

**Comment by nigeljamesstevenson** [comment]
> Thanks for getting this in @nigeljamesstevenson!
> 
> When I went to test this locally, and I wasn't able to get the tests to pass. I kept hitting an invalid username error (which resulted in a 500 returned status for the tests):
> 
> ```
> {
>   "code": "invalid_username",
>   "message": "<strong>Error</strong>: The username <strong>undefined</strong> is not registered on this site. If you are unsure of your username, try your email address instead.",
>   "data": null
> }
> ```
> 
> I am not sure if this was the way I spun up the test environment. I did the following: in `plugins/woocommerce` I ran `pnpm docker:up` (I wasn't able to get `pnpm docker:up --filter=woocommerce` to work for some reason, but this did).
> 
> Is there any other configuration that needs to happen (such as a local `.env` file or something similar) to set things up? Also wanted to check in how you were spinning up and initializing the test environment just to make sure I didn't miss a step, either.
> 
> Thanks!

Hi @zhongruige - thanks for taking a peek! :)

To spin up the environment, I performed the following in `plugins/woocommerce`:

`pnpm env:test --filter=woocommerce`

I also have a .env file in the woocommerce directory.

good shout on the gitignore!

I'll touch base with you on slack re the .env etc to see if we can get things rolling.

**Comment by zhongruige** [comment]
Hi @nigeljamesstevenson!

I'm curious, did we get any benchmarks or comparisons to see runtimes of these tests compared to the original implementation? I'm quire curious if we've seen some nice gains from switching to Playwright's parallelization.

Also it looks like, similar to the Playwright migration for e2e tests, that everything in the `packages/js/api-core-tests` will just contain the original tests we had. Eventually when we go to release a new package with these, we'll want to figure out a migration path for them.

File: plugins/woocommerce/tests/api-core-tests/output/index.html [hunk 1]
```diff
@@ -0,0 +1,62 @@
+
+
+<!DOCTYPE html>
```
Comment (zhongruige) [inline plugins/woocommerce/tests/api-core-tests/output/index.html]: We'll want to delete this file then add this folder to a `.gitignore` file.

File: plugins/woocommerce/tests/api-core-tests/playwright.config.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
+const { devices } = require( '@playwright/test' );
+require( 'dotenv' ).config();
```
Comment (zhongruige) [inline plugins/woocommerce/tests/api-core-tests/playwright.config.js:2]: We chatted on this a bit, but let's see if we can find a way to make this local to this folder as a start. Or, perhaps, we can look at hardcoding it for now as the Playwright e2e tests are doing--then look at making it variable in a future enhancement.

File: plugins/woocommerce/tests/api-core-tests/playwright.config.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+const base64auth = btoa(
```
Comment (zhongruige) [inline plugins/woocommerce/tests/api-core-tests/playwright.config.js]: Since these tests will fail if you don't create the `.env` file in the `plugins/woocommerce` directory (which we likely won't want to create this here in the future anyway), so I had a minor suggestion. Let me know your thoughts on this @nigeljamesstevenson. I am thinking we can update the auth logic to be the following:

```javascript
let userKey = 'admin';
let userSecret = 'password';

if ( process.env.USER_KEY ) {
	userKey = process.env.USER_KEY;
}

if ( process.env.USER_SECRET ) {
	userSecret = process.env.USER_SECRET;
}

const base64auth = btoa( `${ userKey }:${ userSecret }` );
```

This way we can handle if one isn't set. Eventually we'll want to support a local `.env` file to the tests, but we can handle that as we approach preparing to release the package. Otherwise, if we don't do this change, we at least will want to document that you need to have that `.env` file set in the `plugins/woocommerce` directory otherwise the tests won't run. I lean towards adding in the conditional logic above as an initial fix then we can make it dynamic in a future PR.

I made this change locally and tested that it works both with the `.env` file and without it.

**Review by zhongruige — CHANGES_REQUESTED** [review]
Suggested change around the auth logic, otherwise everything looks good to me!

**Comment by zhongruige** [comment]
This tested great for me locally, nice work on this!

Screen Shot 2022-09-30 at 10 40 48

Test report looked great in, too:

Screen Shot 2022-09-30 at 10 41 13

File: plugins/woocommerce/tests/api-core-tests/playwright.config.js (hunk 3 above)
Comment (nigeljamesstevenson) [reply plugins/woocommerce/tests/api-core-tests/playwright.config.js]: This sounds good to me @zhongruige 👍

The original implementation was following the pattern of the existing supertest tests (and therefore using the `.env` file in the `plugins/woocommerce` directory) - I agree the logic can be updated (and make execution one step simpler for the user). 
I will also add a note regarding the use of an `.env` file for completion.. 
I also added in similar for the baseURL to keep it in line with the username and password.

In short - consider it done!

I also noticed that the test execution instructions in the readme didn't clearly state the command to kick off the tests so added an extra line in there and also referenced the fact that the destroy and create test env commands may also need to be executed in `plugins/woocommerce` (or a subdirectory thereof)

**Review by zhongruige — APPROVED** [review]
LGTM! :shipit: 

Thank you for the great work on this @nigeljamesstevenson!

File: plugins/woocommerce/tests/api-core-tests/playwright.config.js (hunk 3 above)
Comment (zhongruige) [reply plugins/woocommerce/tests/api-core-tests/playwright.config.js]: Thank you for making these changes and also for updating the documentation, too!
